add lifted doc (#2061)
karmada-io/karmada: docs/contributors/devel/lifted.md

> # Overview

Reviewer: @Tingtal Please help take a look.

This document explains how lifted code are managed.
A common user case for this task is developer lifting code from other repository to `pkg/util/lifted` directory.

- [Steps of lifting code](#steps-of-lifting-code)
- [How to write lifted comments](#how-to-write-lifted-comments)
- [Examples](#examples)

## Steps of lifting code

> - Copy code from other repository and save it to a go file under `pkg/util/lifted`.

Reviewer 2: ```suggestion
- Copy code from another repository and save it to a go file under `pkg/util/lifted`.
```

- Optionally change the lifted code.

> - Add lifted comments for the code, referring to [it](#how-to-write-liftd-comments).

Reviewer 2: ```suggestion
- Add lifted comments for the code [as guided](#how-to-write-liftd-comments).
```

> - run `hack/update-lifted.sh` to update the lifted doc `pkg/util/lifted/doc.go`

Reviewer 2: ```suggestion
- Run `hack/update-lifted.sh` to update the lifted doc `pkg/util/lifted/doc.go`.
```

## How to write lifted comments

> Lifted comments shall be placed just before the lifted code(may be a func, a type, a var or a const). Except empty line and comments, any others are not allowed between lifted comments and lifted code.

Reviewer 2: ```suggestion
Lifted comments shall be placed just before the lifted code (could be a func, type, var or const). Only empty lines and comments are allowed between lifted comments and lifted code.
```

> Lifted comments are composed by one or multi comment lines. Each line has format: `+lifted:KEY[=VALUE]`. Value is optional for some key. 

Reviewer 2: ```suggestion
Lifted comments are composed by one or multi comment lines, each in the format of `+lifted:KEY[=VALUE]`. Value is optional for some keys. 
```

Valid keys are as follow：

- source: 

  Key `source` is required. Its value indicates where the code is lifted from.

- changed: 

>   Key `changed` is optional. It indicates whether the code is changed. Value is optional (`true` or `false`, defaults to `true`). If `changed` is nonexistent，it's similar to `change=false`, indicating the code is unchanged.

Author: @Tingtal PTAL this line.

Reviewer 2: I would say:

```suggestion
  Key `changed` is optional. It indicates whether the code is changed. Value is optional (`true` or `false`, defaults to `true`). Not adding this key or setting it to `false` means no code change.
```

>   Key `changed` is optional. If not given, means the lifted code is not changed. Otherwise its value indicates. Value of `changed` is optional, the valid value is `true` or `false`, default to be `true` if not set.

Reviewer 2: ```suggestion
  Key `changed` is optional. Without it, the lifted code is not changed. Otherwise its value indicates. Value of `changed` is optional (`true` or `false`, defaults to `true`).
```

Reviewer 2: Something is missing in "Otherwise its value indicates."
There should be something following "indicates".
Pls check.

## Examples
### Lifting function

> Lift function `IsQuotaHugePageResourceName` to `corehelpers.go  `:

Reviewer 2: ```suggestion
Lift function `IsQuotaHugePageResourceName` to `corehelpers.go`:
```

```go
// +lifted:source=https://github.com/kubernetes/kubernetes/blob/release-1.23/pkg/apis/core/helper/helpers.go#L57-L61

// IsQuotaHugePageResourceName returns true if the resource name has the quota
// related huge page resource prefix.
func IsQuotaHugePageResourceName(name corev1.ResourceName) bool {
	return strings.HasPrefix(string(name), corev1.ResourceHugePagesPrefix) || strings.HasPrefix(string(name), corev1.ResourceRequestsHugePagesPrefix)
}
```

Added in `doc.go`:

```markdown
| lifted file              | source file                                                                                                                   | const/var/type/func                     | changed |
|--------------------------|-------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------|---------|
| corehelpers.go           | https://github.com/kubernetes/kubernetes/blob/release-1.23/pkg/apis/core/helper/helpers.go#L57-L61                            | func IsQuotaHugePageResourceName        | N       |
```

### Changed lifting function

> Lift and change function `GetNewReplicaSet` to `deployment.go `

Reviewer 2: ```suggestion
Lift and change function `GetNewReplicaSet` to `deployment.go`
```

| deployment.go            | https://github.com/kubernetes/kubernetes/blob/release-1.22/pkg/controller/deployment/util/deployment_util.go#L536-L544        | func GetNewReplicaSet                   | Y       |
```

### Lifting const

Lift const `isNegativeErrorMsg` to `corevalidation.go  `:

```go
// +lifted:source=https://github.com/kubernetes/kubernetes/blob/release-1.22/pkg/apis/core/validation/validation.go#L59
const isNegativeErrorMsg string = apimachineryvalidation.IsNegativeErrorMsg
```

Added in `doc.go`:

```markdown
| lifted file              | source file                                                                                                                   | const/var/type/func                     | changed |
|--------------------------|-------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------|---------|
| corevalidation.go        | https://github.com/kubernetes/kubernetes/blob/release-1.22/pkg/apis/core/validation/validation.go#L59                         | const isNegativeErrorMsg                | N       |
```

### Lifting type

> Lift type `Visitor` to `visitpod.go  `:

Reviewer 2: ```suggestion
Lift type `Visitor` to `visitpod.go`:
```
